sdk-tools/src/edger8r.rs

```diff
@@ -0,0 +1,374 @@
+    string::{FromUtf8Error, String},
+};
+
+/// Errors which can occur when working with the edger8r tool.
+#[derive(Debug, Display)]
+pub enum Error {
+    /// There was an error running the command: {0}
+    Io(IoError),
+    /// The edger8r command failed, and also printed invalid UTF-8
+    Utf8Error,
+    /**
+     * There was an error generating the code,
+     * command:\n{0}\nstdout:\n{0}\n\nstderr:\n{1}
+     */
+    Generate(String, String, String),
+}
+
+impl From<FromUtf8Error> for Error {
+    fn from(_src: FromUtf8Error) -> Error {
+        Error::Utf8Error
+    }
+}
+
+impl From<IoError> for Error {
+    fn from(src: IoError) -> Error {
+        Error::Io(src)
+    }
+}
+
+/// The kind of C files to generate with the edger8r tool.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Default)]
+pub enum OutputKind {
+    /// Generate code for untrusted
+    Untrusted,
+    /// Generate code for trusted
+    Trusted,
+    /// Generate code for both trusted and untrusted
+    #[default]
+    Both,
+}
+
+/// Wrapper for the edger8r tool.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Edger8r {
+    /// The path to the `sgx_edger8r` executable.
+    edger8r_path: PathBuf,
+    /// The build output directory.
+    out_dir: Option<PathBuf>,
+    /// The path to the EDL file to generate code for.
+    edl_file: PathBuf,
+    /// The type of code to be generated.
+    output_kind: OutputKind,
+}
+
+/// The files generated by the edger8r tool
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Edger8rFiles {
+    /// The trusted files generated. Will be empty when
+    /// [`OutputKind::Untrusted`] is used.
+    pub trusted: Vec<PathBuf>,
+    /// The untrusted files generated. Will be empty when
+    /// [`OutputKind::Trusted`] is used.
+    pub untrusted: Vec<PathBuf>,
+}
+
+impl Edger8r {
+    /// Create a new edger8r builder.
+    ///
+    /// # Arguments
+    /// * `edl_file` - The edl file to generate C files from
+    pub fn new(edl_file: impl AsRef<Path>) -> Self {
+        let edl_file = edl_file.as_ref().to_owned();
+
+        Self {
+            edger8r_path: Self::edger8r_path(),
+            out_dir: None,
+            edl_file,
+            output_kind: OutputKind::Both,
+        }
+    }
+
+    /// Set the type of code we should generate
+    ///
+    /// # Arguments
+    /// * `kind` - The kind of code to generate
+    pub fn output_kind(&mut self, kind: OutputKind) -> &mut Self {
+        self.output_kind = kind;
+        self
+    }
+
+    /// Specify the output directory
+    ///
+    /// # Arguments
+    /// * `out_dir` - the output directory to place the generated files in.
+    pub fn out_dir(&mut self, out_dir: impl AsRef<Path>) -> &mut Self {
+        self.out_dir = Some(out_dir.as_ref().to_owned());
+        self
+    }
+
+    /// Generate C files using the edger8r tool
+    ///
+    /// # Returns:
+    /// The paths to the generated files.
+    pub fn generate(&self) -> Result<Edger8rFiles, Error> {
```

Comment:
in the [MobileCoin repo](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/util/build/sgx/src/edger8r.rs#L162) this returned `Result<&Self, Error>` so that one could do
`edger8r.generate().build()`. To minimize the size of this initial PR I only brought in the `generate()` and it felt awkward returning self from it.  I'm not sure if this, `Edger8r` struct should have the logic to compile the files that seems like general C code compilation and not specific to edger8r generated code.

Reply:
I'm fine to move `build()` out, but then this builder should be changed to consuming (i.e. `foo(self) -> Self` instead of `foo(&self) -> &Self`).

Reply:
updated `generate()` to be consuming now